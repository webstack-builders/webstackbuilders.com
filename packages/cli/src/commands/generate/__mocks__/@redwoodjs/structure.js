// We need this because we check for typescript automatically in generate

const mockedStructure = {
  ...jest.requireActual('@redwoodjs/structure'),
  getProject: () => {
    return {
      isTypeScriptProject: false,
    }
  },
}

module.exports = mockedStructure
