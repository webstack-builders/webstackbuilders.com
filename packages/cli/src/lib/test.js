/* eslint-env jest */

// Include at the top of your tests. Automatically mocks out the file system
//
// import { loadComponentFixture } from 'src/lib/test'
//
// test('true is true', () => {
//   expect('some output').toEqual(loadComponentFixture('component', 'filename.js'))
// })

import fs from 'fs'
import path from 'path'

jest.mock('@redwoodjs/internal', () => {
  const path = require('path')
  return {
    ...jest.requireActual('@redwoodjs/internal'),
    generate: () => {},
    getPaths: () => {
      const BASE_PATH = '/path/to/project'
      return {
        base: BASE_PATH,
        api: {
          dataMigrations: path.join(BASE_PATH, './api/prisma/dataMigrations'),
          db: path.join(global.__dirname, 'fixtures'), // this folder
          src: path.join(BASE_PATH, './api/src'),
          services: path.join(BASE_PATH, './api/src/services'),
          graphql: path.join(BASE_PATH, './api/src/graphql'),
          functions: path.join(BASE_PATH, './api/src/functions'),
        },
        web: {
          src: path.join(BASE_PATH, './web/src'),
          routes: path.join(BASE_PATH, 'web/src/Routes.js'),
          components: path.join(BASE_PATH, '/web/src/components'),
          layouts: path.join(BASE_PATH, '/web/src/layouts'),
          pages: path.join(BASE_PATH, '/web/src/pages'),
        },
        scripts: path.join(BASE_PATH, 'scripts'),
        generated: {
          base: path.join(BASE_PATH, '.redwood'),
          schema: path.join(BASE_PATH, '.redwood/schema.graphql'),
          types: {
            includes: path.join(BASE_PATH, '.redwood/types/includes'),
            mirror: path.join(BASE_PATH, '.redwood/types/mirror'),
          },
        },
      }
    },
  }
})

global.__prettierPath = path.resolve(
  __dirname,
  './__tests__/fixtures/prettier.config.js'
)

jest.mock('path', () => {
  const path = jest.requireActual('path')
  return {
    ...path,
    join: (...paths) => {
      if (
        paths &&
        paths[0] === '/path/to/project' &&
        paths[1] === 'prettier.config.js'
      ) {
        return global.__prettierPath
      }
      return path.join(...paths)
    },
  }
})

export const generatorsRootPath = path.join(
  __dirname,
  '..',
  'commands',
  'generate'
)

/**
 * Loads the fixture for a generator by assuming a lot of the path structure
 * automatically:
 *
 *   `loadGeneratorFixture('scaffold', 'NamePage.js')`
 *
 * will return the contents of:
 *
 *   `cli/src/commands/generate/scaffold/__tests__/fixtures/NamePage.js`
 */
export const loadGeneratorFixture = (generator, name) => {
  return loadFixture(
    path.join(
      __dirname,
      '..',
      'commands',
      'generate',
      generator,
      '__tests__',
      'fixtures',
      name
    )
  )
}

/**
 * Returns the contents of a text file in a `fixtures` directory
 */
export const loadFixture = (filepath) => {
  return fs.readFileSync(filepath).toString()
}
