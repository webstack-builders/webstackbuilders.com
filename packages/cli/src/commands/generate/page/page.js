import { execSync } from 'child_process'

import camelcase from 'camelcase'
import Listr from 'listr'
import pascalcase from 'pascalcase'

import { getConfig, generate as generateTypes } from '@redwoodjs/internal'

import {
  addRoutesToRouterTask,
  transformTSToJS,
  writeFilesTask,
} from '../../../lib'
import c from '../../../lib/colors'
import {
  createYargsForComponentGeneration,
  pathName,
  templateForComponentFile,
  mapRouteParamTypeToTsType,
} from '../helpers'

const COMPONENT_SUFFIX = 'Page'
const REDWOOD_WEB_PATH_NAME = 'pages'

/** @type {(paramType: 'Int' | 'Boolean' | 'String') } **/
const mapRouteParamTypeToDefaultValue = (paramType) => {
  switch (paramType) {
    case 'Int':
      return 42

    case 'Float':
      return 42.1

    case 'Boolean':
      return true

    default:
      // Boolean -> boolean, String -> string
      return '42'
  }
}

export const paramVariants = (path) => {
  const param = path?.match(/(\{[\w:]+\})/)?.[1]
  const paramName = param?.replace(/:[^}]+/, '').slice(1, -1)

  if (param === undefined) {
    return {
      propParam: '',
      propValueParam: '',
      argumentParam: '',
      paramName: '',
      paramValue: '',
      paramType: '',
    }
  }

  // set paramType param includes type (e.g. {id:Int}), else use string
  const routeParamType = param?.match(/:/)
    ? param?.replace(/[^:]+/, '').slice(1, -1)
    : 'String'

  const defaultValue = mapRouteParamTypeToDefaultValue(routeParamType)
  const defaultValueAsProp =
    routeParamType === 'String' ? `'${defaultValue}'` : defaultValue

  // "42" is just a value used for demonstrating parameter usage in the
  // generated page-, test-, and story-files.
  return {
    propParam: `{ ${paramName} }`,
    propValueParam: `${paramName}={${defaultValueAsProp}} `, // used it story
    argumentParam: `{ ${paramName}: ${defaultValueAsProp} }`,
    paramName,
    paramValue: defaultValue,
    paramType: mapRouteParamTypeToTsType(routeParamType),
  }
}

export const files = ({ name, tests, stories, typescript, ...rest }) => {
  const pageFile = templateForComponentFile({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: typescript ? '.tsx' : '.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'page',
    templatePath: 'page.tsx.template',
    templateVars: rest,
  })

  const testFile = templateForComponentFile({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: typescript ? '.test.tsx' : '.test.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'page',
    templatePath: 'test.tsx.template',
    templateVars: rest,
  })

  const storiesFile = templateForComponentFile({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: typescript ? '.stories.tsx' : '.stories.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'page',
    templatePath: 'stories.tsx.template',
    templateVars: rest,
  })

  const files = [pageFile]

  if (tests) {
    files.push(testFile)
  }

  if (stories) {
    files.push(storiesFile)
  }

  // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }
  return files.reduce((acc, [outputPath, content]) => {
    const template = typescript ? content : transformTSToJS(outputPath, content)

    return {
      [outputPath]: template,
      ...acc,
    }
  }, {})
}

export const routes = ({ name, path }) => {
  return [
    `<Route path="${path}" page={${pascalcase(name)}Page} name="${camelcase(
      name
    )}" />`,
  ]
}

const positionalsObj = {
  path: {
    description: 'URL path to the page, or just {param}. Defaults to name',
    type: 'string',
  },
}

// @NOTE: Not exporting handler from function
// As pages need a special handler
export const { command, description, builder } =
  createYargsForComponentGeneration({
    componentName: 'page',
    filesFn: files,
    positionalsObj,
  })

export const handler = async ({
  name,
  path,
  force,
  tests,
  stories,
  typescript = false,
}) => {
  if (tests === undefined) {
    tests = getConfig().generate.tests
  }
  if (stories === undefined) {
    stories = getConfig().generate.stories
  }
  if (process.platform === 'win32') {
    // running `yarn rw g page home /` on Windows using GitBash
    // POSIX-to-Windows path conversion will kick in.
    // See https://github.com/git-for-windows/build-extra/blob/d715c9e/ReleaseNotes.md
    // As a workaround we try to detect when this has happened, and reverse
    // the action

    try {
      // `cygpath -m /` will return something like 'C:/Program Files/Git/\n'
      const slashPath = execSync('cygpath -m /', {
        stdio: ['ignore', 'pipe', 'ignore'],
      })
        .toString()
        .trim()

      // `yarn rw g page home /` =>
      //   page === 'C:/Program Files/Git'
      // `yarn rw g page about /about` =>
      //   page === 'C:/Program Files/Git/about'
      // Sometimes there is a / after 'Git' to match, sometimes there isn't
      path = path.replace(new RegExp(`^${slashPath}?`), '/')
    } catch {
      // probably using PowerShell or cmd, in which case no special handling
      // is needed
    }
  }

  const tasks = new Listr(
    [
      {
        title: 'Generating page files...',
        task: async () => {
          path = pathName(path, name)
          const f = await files({
            name,
            path,
            tests,
            stories,
            typescript,
            ...paramVariants(path),
          })
          return writeFilesTask(f, { overwriteExisting: force })
        },
      },
      {
        title: 'Updating routes file...',
        task: async () => {
          addRoutesToRouterTask(routes({ name, path: pathName(path, name) }))
        },
      },
      {
        title: `Generating types ...`,
        task: async () => {
          return generateTypes()
        },
      },
    ].filter(Boolean),
    { collapse: false }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.error(c.error(e.message))
    process.exit(e?.exitCode || 1)
  }
}
