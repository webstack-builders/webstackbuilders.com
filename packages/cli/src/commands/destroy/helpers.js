import Listr from 'listr'

import { deleteFilesTask } from '../../lib'
import c from '../../lib/colors'
import { ensureUniquePlural } from '../generate/helpers'

const tasks = ({ componentName, filesFn, name }) =>
  new Listr(
    [
      {
        title: `Destroying ${componentName} files...`,
        task: async () => {
          const f = await filesFn({ name, stories: true, tests: true })
          return deleteFilesTask(f)
        },
      },
    ],
    { collapse: false, exitOnError: true }
  )

export const createYargsForComponentDestroy = ({
  componentName,
  filesFn,
  shouldEnsureUniquePlural = false,
}) => {
  return {
    command: `${componentName} <name>`,
    description: `Destroy a ${componentName} component`,
    builder: (yargs) => {
      yargs.positional('name', {
        description: `Name of the ${componentName}`,
        type: 'string',
      })
    },
    handler: async ({ name }) => {
      if (shouldEnsureUniquePlural) {
        await ensureUniquePlural({ model: name, inDestroyer: true })
      }
      const t = tasks({ componentName, filesFn, name })

      try {
        await t.run()
      } catch (e) {
        console.log(c.error(e.message))
      }
    },
    tasks,
  }
}
