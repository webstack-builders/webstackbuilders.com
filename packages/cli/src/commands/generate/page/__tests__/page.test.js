global.__dirname = __dirname

global.mockFs = false
let mockFiles = {}

jest.mock('fs', () => {
  const actual = jest.requireActual('fs')

  return {
    ...actual,
    existsSync: (...args) => {
      if (!global.mockFs) {
        return actual.existsSync.apply(null, args)
      }
      return false
    },
    mkdirSync: (...args) => {
      if (!global.mockFs) {
        return actual.mkdirSync.apply(null, args)
      }
    },
    writeFileSync: (target, contents) => {
      if (!global.mockFs) {
        return actual.writeFileSync.call(null, target, contents)
      }
    },
    readFileSync: (path) => {
      if (!global.mockFs) {
        return actual.readFileSync.call(null, path)
      }

      const mockedContent = mockFiles[path]

      return mockedContent || actual.readFileSync.call(null, path)
    },
  }
})

import fs from 'fs'
import path from 'path'

// Load mocks
import '../../../../lib/test'

import { ensurePosixPath } from '@redwoodjs/internal'

import { getPaths } from '../../../../lib'
import { pathName } from '../../helpers'
import * as page from '../page'

let singleWordFiles,
  multiWordFiles,
  pluralWordFiles,
  paramFiles,
  noTestsFiles,
  noStoriesFiles,
  typescriptFiles,
  typescriptParamFiles,
  typescriptParamTypeFiles

beforeAll(() => {
  singleWordFiles = page.files({
    name: 'Home',
    tests: true,
    stories: true,
    ...page.paramVariants(pathName(undefined, 'home')),
  })
  multiWordFiles = page.files({
    name: 'ContactUs',
    tests: true,
    stories: true,
    ...page.paramVariants(pathName(undefined, 'contact-us')),
  })
  pluralWordFiles = page.files({
    name: 'Cats',
    tests: true,
    stories: true,
    ...page.paramVariants(pathName(undefined, 'cats')),
  })
  paramFiles = page.files({
    name: 'Post',
    tests: true,
    stories: true,
    ...page.paramVariants(pathName('{id}', 'post')),
  })
  noTestsFiles = page.files({
    name: 'NoTests',
    tests: false,
    stories: true,
    ...page.paramVariants(pathName(undefined, 'no-tests')),
  })
  noStoriesFiles = page.files({
    name: 'NoStories',
    tests: true,
    stories: false,
    ...page.paramVariants(pathName(undefined, 'no-stories')),
  })
  typescriptFiles = page.files({
    name: 'TSFiles',
    typescript: true,
    tests: true,
    stories: true,
    ...page.paramVariants(pathName(undefined, 'typescript')),
  })
  typescriptParamFiles = page.files({
    name: 'TSParamFiles',
    typescript: true,
    tests: true,
    stories: true,
    ...page.paramVariants(pathName('{id}', 'typescript-param')),
  })
  typescriptParamTypeFiles = page.files({
    name: 'TSParamTypeFiles',
    typescript: true,
    tests: true,
    stories: true,
    ...page.paramVariants(pathName('{id:Int}', 'typescript-param-with-type')),
  })
})

test('returns exactly 3 files', () => {
  expect(Object.keys(singleWordFiles).length).toEqual(3)
})

test('creates a page component', () => {
  expect(
    singleWordFiles[
      path.normalize('/path/to/project/web/src/pages/HomePage/HomePage.js')
    ]
  ).toMatchSnapshot()
})

test('creates a page test', () => {
  expect(
    singleWordFiles[
      path.normalize('/path/to/project/web/src/pages/HomePage/HomePage.test.js')
    ]
  ).toMatchSnapshot()
})

test('creates a page story', () => {
  expect(
    singleWordFiles[
      path.normalize(
        '/path/to/project/web/src/pages/HomePage/HomePage.stories.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a page component', () => {
  expect(
    multiWordFiles[
      path.normalize(
        '/path/to/project/web/src/pages/ContactUsPage/ContactUsPage.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a test for a component with multiple words for a name', () => {
  expect(
    multiWordFiles[
      path.normalize(
        '/path/to/project/web/src/pages/ContactUsPage/ContactUsPage.test.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a page story', () => {
  expect(
    multiWordFiles[
      path.normalize(
        '/path/to/project/web/src/pages/ContactUsPage/ContactUsPage.stories.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a page component with a plural word for name', () => {
  expect(
    pluralWordFiles[
      path.normalize('/path/to/project/web/src/pages/CatsPage/CatsPage.js')
    ]
  ).toMatchSnapshot()
})

test('creates a page component with params', () => {
  expect(
    paramFiles[
      path.normalize('/path/to/project/web/src/pages/PostPage/PostPage.js')
    ]
  ).toMatchSnapshot()
})

test('creates a test for page component with params', () => {
  expect(
    paramFiles[
      path.normalize('/path/to/project/web/src/pages/PostPage/PostPage.test.js')
    ]
  ).toMatchSnapshot()
})

test('doesnt create a test for page component when tests=false', () => {
  expect(Object.keys(noTestsFiles)).toEqual([
    path.normalize(
      '/path/to/project/web/src/pages/NoTestsPage/NoTestsPage.stories.js'
    ),
    path.normalize('/path/to/project/web/src/pages/NoTestsPage/NoTestsPage.js'),
  ])
})

test('doesnt create a story for page component when stories=false', () => {
  expect(Object.keys(noStoriesFiles)).toEqual([
    path.normalize(
      '/path/to/project/web/src/pages/NoStoriesPage/NoStoriesPage.test.js'
    ),
    path.normalize(
      '/path/to/project/web/src/pages/NoStoriesPage/NoStoriesPage.js'
    ),
  ])
})

test('creates a single-word route name', () => {
  const names = ['Home', 'home']

  names.forEach((name) => {
    expect(page.routes({ name: name, path: '/' })).toEqual([
      '<Route path="/" page={HomePage} name="home" />',
    ])
  })
})

test('creates a camelCase route name for multiple word names', () => {
  const names = ['FooBar', 'foo_bar', 'foo-bar', 'fooBar']

  names.forEach((name) => {
    expect(page.routes({ name: name, path: 'foo-bar' })).toEqual([
      '<Route path="foo-bar" page={FooBarPage} name="fooBar" />',
    ])
  })
})

test('creates a path equal to passed path', () => {
  expect(page.routes({ name: 'FooBar', path: 'fooBar-baz' })).toEqual([
    '<Route path="fooBar-baz" page={FooBarPage} name="fooBar" />',
  ])
})

test('paramVariants returns empty strings for no params', () => {
  const emptyParams = {
    propParam: '',
    propValueParam: '',
    argumentParam: '',
    paramName: '',
    paramValue: '',
    paramType: '',
  }
  expect(page.paramVariants()).toEqual(emptyParams)
  expect(page.paramVariants('')).toEqual(emptyParams)
  expect(page.paramVariants('/')).toEqual(emptyParams)
  expect(page.paramVariants('/post/edit')).toEqual(emptyParams)
})

test('paramVariants finds the param and type in the middle of the path', () => {
  expect(page.paramVariants('/post/{id:Int}/edit')).toEqual({
    propParam: '{ id }',
    propValueParam: 'id={42} ',
    argumentParam: '{ id: 42 }',
    paramName: 'id',
    paramValue: 42,
    paramType: 'number',
  })
})

test('paramVariants handles boolean type', () => {
  expect(page.paramVariants('/post/edit/{debug:Boolean}')).toEqual({
    propParam: '{ debug }',
    propValueParam: 'debug={true} ',
    argumentParam: '{ debug: true }',
    paramName: 'debug',
    paramValue: true,
    paramType: 'boolean',
  })
})

test('paramVariants paramType defaults to string', () => {
  expect(page.paramVariants('/posts/{id}')).toEqual({
    propParam: '{ id }',
    propValueParam: "id={'42'} ",
    argumentParam: "{ id: '42' }",
    paramName: 'id',
    paramValue: '42',
    paramType: 'string',
  })
})

test('file generation', async () => {
  mockFiles = {
    [getPaths().web.routes]: [
      "import { Router, Route } from '@redwoodjs/router'",
      '',
      'const Routes = () => {',
      '  return (',
      '    <Router>',
      '      <Route path="/about" page={AboutPage} name="about" />',
      '      <Route notfound page={NotFoundPage} />',
      '    </Router>',
      '  )',
      '}',
      '',
      'export default Routes',
    ].join('\n'),
  }

  const spy = jest.spyOn(fs, 'writeFileSync')

  global.mockFs = true

  await page.handler({
    name: 'home',
    path: '',
    force: false,
    tests: true,
    stories: true,
  })

  expect(spy).toHaveBeenCalled()

  spy.mock.calls.forEach((calls) => {
    const testOutput = {
      // Because windows paths are different, we need to normalise before snapshotting
      filePath: ensurePosixPath(calls[0]),
      fileContent: calls[1],
    }
    expect(testOutput).toMatchSnapshot()
  })

  global.mockFs = false
  spy.mockRestore()
})

test('file generation with route params', async () => {
  mockFiles = {
    [getPaths().web.routes]: [
      "import { Router, Route } from '@redwoodjs/router'",
      '',
      'const Routes = () => {',
      '  return (',
      '    <Router>',
      '      <Route path="/about" page={AboutPage} name="about" />',
      '      <Route notfound page={NotFoundPage} />',
      '    </Router>',
      '  )',
      '}',
      '',
      'export default Routes',
    ].join('\n'),
  }

  const spy = jest.spyOn(fs, 'writeFileSync')
  global.mockFs = true

  await page.handler({
    name: 'post',
    path: '{id}',
    force: false,
    tests: true,
    stories: true,
  })

  expect(spy).toHaveBeenCalled()

  spy.mock.calls.forEach((calls) => {
    const testOutput = {
      filePath: ensurePosixPath(calls[0]),
      fileContent: calls[1],
    }
    expect(testOutput).toMatchSnapshot()
  })

  global.mockFs = false
  spy.mockRestore()
})

test('generates typescript pages', () => {
  expect(
    typescriptFiles[
      path.normalize(
        '/path/to/project/web/src/pages/TsFilesPage/TsFilesPage.tsx'
      )
    ]
  ).toMatchSnapshot()

  expect(
    typescriptFiles[
      path.normalize(
        '/path/to/project/web/src/pages/TsFilesPage/TsFilesPage.stories.tsx'
      )
    ]
  ).toMatchSnapshot()

  expect(
    typescriptFiles[
      path.normalize(
        '/path/to/project/web/src/pages/TsFilesPage/TsFilesPage.test.tsx'
      )
    ]
  ).toMatchSnapshot()

  expect(
    typescriptParamFiles[
      path.normalize(
        '/path/to/project/web/src/pages/TsParamFilesPage/TsParamFilesPage.tsx'
      )
    ]
  ).toMatchSnapshot()

  expect(
    typescriptParamTypeFiles[
      path.normalize(
        '/path/to/project/web/src/pages/TsParamTypeFilesPage/TsParamTypeFilesPage.tsx'
      )
    ]
  ).toMatchSnapshot()
})
