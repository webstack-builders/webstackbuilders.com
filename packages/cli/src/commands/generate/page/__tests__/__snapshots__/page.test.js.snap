// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a page component 1`] = `
"import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

const HomePage = () => {
  return (
    <>
      <MetaTags
        title=\\"Home\\"
        // description=\\"Home description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
      You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />

      <h1>HomePage</h1>
      <p>
        Find me in <code>./web/src/pages/HomePage/HomePage.js</code>
      </p>
      <p>
        My default route is named <code>home</code>, link to me with \`
        <Link to={routes.home()}>Home</Link>\`
      </p>
    </>
  )
}

export default HomePage
"
`;

exports[`creates a page component 2`] = `
"import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

const ContactUsPage = () => {
  return (
    <>
      <MetaTags
        title=\\"ContactUs\\"
        // description=\\"ContactUs description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
      You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />

      <h1>ContactUsPage</h1>
      <p>
        Find me in <code>./web/src/pages/ContactUsPage/ContactUsPage.js</code>
      </p>
      <p>
        My default route is named <code>contactUs</code>, link to me with \`
        <Link to={routes.contactUs()}>ContactUs</Link>\`
      </p>
    </>
  )
}

export default ContactUsPage
"
`;

exports[`creates a page component with a plural word for name 1`] = `
"import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

const CatsPage = () => {
  return (
    <>
      <MetaTags
        title=\\"Cats\\"
        // description=\\"Cats description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
      You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />

      <h1>CatsPage</h1>
      <p>
        Find me in <code>./web/src/pages/CatsPage/CatsPage.js</code>
      </p>
      <p>
        My default route is named <code>cats</code>, link to me with \`
        <Link to={routes.cats()}>Cats</Link>\`
      </p>
    </>
  )
}

export default CatsPage
"
`;

exports[`creates a page component with params 1`] = `
"import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

const PostPage = ({ id }) => {
  return (
    <>
      <MetaTags
        title=\\"Post\\"
        // description=\\"Post description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
      You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />

      <h1>PostPage</h1>
      <p>
        Find me in <code>./web/src/pages/PostPage/PostPage.js</code>
      </p>
      <p>
        My default route is named <code>post</code>, link to me with \`
        <Link to={routes.post({ id: '42' })}>Post 42</Link>\`
      </p>
      <p>The parameter passed to me is {id}</p>
    </>
  )
}

export default PostPage
"
`;

exports[`creates a page story 1`] = `
"import HomePage from './HomePage'

export const generated = () => {
  return <HomePage />
}

export default { title: 'Pages/HomePage' }
"
`;

exports[`creates a page story 2`] = `
"import ContactUsPage from './ContactUsPage'

export const generated = () => {
  return <ContactUsPage />
}

export default { title: 'Pages/ContactUsPage' }
"
`;

exports[`creates a page test 1`] = `
"import { render } from '@redwoodjs/testing/web'

import HomePage from './HomePage'

describe('HomePage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<HomePage />)
    }).not.toThrow()
  })
})
"
`;

exports[`creates a test for a component with multiple words for a name 1`] = `
"import { render } from '@redwoodjs/testing/web'

import ContactUsPage from './ContactUsPage'

describe('ContactUsPage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<ContactUsPage />)
    }).not.toThrow()
  })
})
"
`;

exports[`creates a test for page component with params 1`] = `
"import { render } from '@redwoodjs/testing/web'

import PostPage from './PostPage'

describe('PostPage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<PostPage id={'42'} />)
    }).not.toThrow()
  })
})
"
`;

exports[`file generation 1`] = `
Object {
  "fileContent": "import HomePage from './HomePage'

export const generated = () => {
  return <HomePage />
}

export default { title: 'Pages/HomePage' }
",
  "filePath": "/path/to/project/web/src/pages/HomePage/HomePage.stories.js",
}
`;

exports[`file generation 2`] = `
Object {
  "fileContent": "import { render } from '@redwoodjs/testing/web'

import HomePage from './HomePage'

describe('HomePage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<HomePage />)
    }).not.toThrow()
  })
})
",
  "filePath": "/path/to/project/web/src/pages/HomePage/HomePage.test.js",
}
`;

exports[`file generation 3`] = `
Object {
  "fileContent": "import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

const HomePage = () => {
  return (
    <>
      <MetaTags
        title=\\"Home\\"
        // description=\\"Home description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
      You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />

      <h1>HomePage</h1>
      <p>
        Find me in <code>./web/src/pages/HomePage/HomePage.js</code>
      </p>
      <p>
        My default route is named <code>home</code>, link to me with \`
        <Link to={routes.home()}>Home</Link>\`
      </p>
    </>
  )
}

export default HomePage
",
  "filePath": "/path/to/project/web/src/pages/HomePage/HomePage.js",
}
`;

exports[`file generation 4`] = `
Object {
  "fileContent": "import { Router, Route } from '@redwoodjs/router'

const Routes = () => {
  return (
    <Router>
      <Route path=\\"/home\\" page={HomePage} name=\\"home\\" />
      <Route path=\\"/about\\" page={AboutPage} name=\\"about\\" />
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes",
  "filePath": "/path/to/project/web/src/Routes.js",
}
`;

exports[`file generation with route params 1`] = `
Object {
  "fileContent": "import PostPage from './PostPage'

export const generated = () => {
  return <PostPage id={'42'} />
}

export default { title: 'Pages/PostPage' }
",
  "filePath": "/path/to/project/web/src/pages/PostPage/PostPage.stories.js",
}
`;

exports[`file generation with route params 2`] = `
Object {
  "fileContent": "import { render } from '@redwoodjs/testing/web'

import PostPage from './PostPage'

describe('PostPage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<PostPage id={'42'} />)
    }).not.toThrow()
  })
})
",
  "filePath": "/path/to/project/web/src/pages/PostPage/PostPage.test.js",
}
`;

exports[`file generation with route params 3`] = `
Object {
  "fileContent": "import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

const PostPage = ({ id }) => {
  return (
    <>
      <MetaTags
        title=\\"Post\\"
        // description=\\"Post description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
      You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />

      <h1>PostPage</h1>
      <p>
        Find me in <code>./web/src/pages/PostPage/PostPage.js</code>
      </p>
      <p>
        My default route is named <code>post</code>, link to me with \`
        <Link to={routes.post({ id: '42' })}>Post 42</Link>\`
      </p>
      <p>The parameter passed to me is {id}</p>
    </>
  )
}

export default PostPage
",
  "filePath": "/path/to/project/web/src/pages/PostPage/PostPage.js",
}
`;

exports[`file generation with route params 4`] = `
Object {
  "fileContent": "import { Router, Route } from '@redwoodjs/router'

const Routes = () => {
  return (
    <Router>
      <Route path=\\"/post/{id}\\" page={PostPage} name=\\"post\\" />
      <Route path=\\"/about\\" page={AboutPage} name=\\"about\\" />
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes",
  "filePath": "/path/to/project/web/src/Routes.js",
}
`;

exports[`generates typescript pages 1`] = `
"import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

const TSFilesPage = () => {
  return (
    <>
      <MetaTags
        title=\\"TSFiles\\"
        // description=\\"TSFiles description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
        You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />
      <h1>TSFilesPage</h1>
      <p>
        Find me in <code>./web/src/pages/TsFilesPage/TsFilesPage.tsx</code>
      </p>
      <p>
        My default route is named <code>tsFiles</code>, link to me with \`
        <Link to={routes.tsFiles()}>TSFiles</Link>\`
      </p>
    </>
  )
}

export default TSFilesPage
"
`;

exports[`generates typescript pages 2`] = `
"import TSFilesPage from './TSFilesPage'

export const generated = () => {
  return <TSFilesPage />
}

export default { title: 'Pages/TSFilesPage' }
"
`;

exports[`generates typescript pages 3`] = `
"import { render } from '@redwoodjs/testing/web'

import TSFilesPage from './TSFilesPage'

describe('TSFilesPage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<TSFilesPage />)
    }).not.toThrow()
  })
})
"
`;

exports[`generates typescript pages 4`] = `
"import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

type TSParamFilesPageProps = {
  id: string
}

const TSParamFilesPage = ({ id }: TSParamFilesPageProps) => {
  return (
    <>
      <MetaTags
        title=\\"TSParamFiles\\"
        // description=\\"TSParamFiles description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
        You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />
      <h1>TSParamFilesPage</h1>
      <p>
        Find me in <code>./web/src/pages/TsParamFilesPage/TsParamFilesPage.tsx</code>
      </p>
      <p>
        My default route is named <code>tsParamFiles</code>, link to me with \`
        <Link to={routes.tsParamFiles({ id: '42' })}>TSParamFiles 42</Link>\`
      </p>
      <p>The parameter passed to me is {id}</p>
    </>
  )
}

export default TSParamFilesPage
"
`;

exports[`generates typescript pages 5`] = `
"import { Link, routes } from '@redwoodjs/router'
import { MetaTags } from '@redwoodjs/web'

type TSParamTypeFilesPageProps = {
  id: number
}

const TSParamTypeFilesPage = ({ id }: TSParamTypeFilesPageProps) => {
  return (
    <>
      <MetaTags
        title=\\"TSParamTypeFiles\\"
        // description=\\"TSParamTypeFiles description\\"
        /* you should un-comment description and add a unique description, 155 characters or less
        You can look at this documentation for best practices : https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets */
      />
      <h1>TSParamTypeFilesPage</h1>
      <p>
        Find me in <code>./web/src/pages/TsParamTypeFilesPage/TsParamTypeFilesPage.tsx</code>
      </p>
      <p>
        My default route is named <code>tsParamTypeFiles</code>, link to me with \`
        <Link to={routes.tsParamTypeFiles({ id: 42 })}>TSParamTypeFiles 42</Link>\`
      </p>
      <p>The parameter passed to me is {id}</p>
    </>
  )
}

export default TSParamTypeFilesPage
"
`;
